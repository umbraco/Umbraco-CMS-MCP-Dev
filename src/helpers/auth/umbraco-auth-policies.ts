import { CurrentUserResponseModel } from "@/umb-management-api/schemas/index.js"

// matches hardcoded value in Umbraco source code
export const AdminGroupKeyString = "E5E7F6C8-7F9C-4B5B-8D5D-9E1E5A4F7E4D";

export const sections = {
    content: "Umb.Section.Content",
    forms: "Umb.Section.Forms",
    media: "Umb.Section.Media",
    members: "Umb.Section.Members",
    packages: "Umb.Section.Packages",
    settings: "Umb.Section.Settings",
    translation: "Umb.Section.Translation",
    workflow: "Umb.Section.Workflow",
    users: "Umb.Section.Users"
}

export const AuthorizationPolicies = {
    RequireAdminAccess: (user: CurrentUserResponseModel) =>
        user.userGroupIds.some((groupId) => groupId.id.toUpperCase() === AdminGroupKeyString),

    SectionAccessContent: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.content),

    SectionAccessContentOrMedia: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.content
            || section === sections.media),

    SectionAccessForContentTree: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.content
            || section === sections.media
            || section === sections.users
            || section === sections.settings
            || section === sections.packages
            || section === sections.members),

    SectionAccessForMediaTree: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.content
            || section === sections.media
            || section === sections.users
            || section === sections.settings
            || section === sections.packages
            || section === sections.members),

    SectionAccessForMemberTree: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.content
            || section === sections.media
            || section === sections.members),

    SectionAccessMedia: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.media),

    SectionAccessMembers: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.members),

    SectionAccessPackages: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.packages),

    SectionAccessSettings: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings),

    SectionAccessUsers: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.users),

    TreeAccessDataTypes: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings),

    TreeAccessDictionary: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.translation),

    TreeAccessDictionaryOrTemplates: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.translation
            || section === sections.settings),

    TreeAccessDocuments: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.content),

    TreeAccessDocumentsOrDocumentTypes: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.content
            || section === sections.settings),

    TreeAccessDocumentOrMediaOrContentTypes: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.content
            || section === sections.settings
            || section === sections.media),

    TreeAccessDocumentsOrMediaOrMembersOrContentTypes: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.content
            || section === sections.settings
            || section === sections.media
            || section === sections.members),

    TreeAccessDocumentTypes: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings),

    TreeAccessLanguages: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings),

    TreeAccessMediaTypes: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings),

    TreeAccessMediaOrMediaTypes: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.media
            || section === sections.settings),

    TreeAccessMemberGroups: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.members),

    TreeAccessMemberTypes: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings),

    TreeAccessMembersOrMemberTypes: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings
            || section === sections.members),

    TreeAccessPartialViews: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings),

    TreeAccessRelationTypes: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings),

    TreeAccessScripts: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings),

    TreeAccessStylesheets: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings),

    TreeAccessStylesheetsOrDocumentOrMediaOrMember: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings
            || section === sections.content
            || section === sections.media
            || section === sections.members),

    TreeAccessTemplates: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings
            || section === sections.content),
    TreeAccessWebhooks: (user: CurrentUserResponseModel) =>
        user.allowedSections.some((section) => section === sections.settings),
};