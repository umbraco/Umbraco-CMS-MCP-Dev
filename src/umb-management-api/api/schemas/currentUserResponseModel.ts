/**
 * Generated by orval v7.8.0 üç∫
 * Do not edit manually.
 * Umbraco Management API
 * This shows all APIs available in this version of Umbraco - including all the legacy apis that are available for backward compatibility
 * OpenAPI spec version: Latest
 */
import type { ReferenceByIdModel } from './referenceByIdModel.js';
import type { CurrentUserResponseModelPermissionsItem } from './currentUserResponseModelPermissionsItem.js';

export interface CurrentUserResponseModel {
  email: string;
  userName: string;
  name: string;
  userGroupIds: ReferenceByIdModel[];
  id: string;
  /** @nullable */
  languageIsoCode: string | null;
  documentStartNodeIds: ReferenceByIdModel[];
  hasDocumentRootAccess: boolean;
  mediaStartNodeIds: ReferenceByIdModel[];
  hasMediaRootAccess: boolean;
  avatarUrls: string[];
  languages: string[];
  hasAccessToAllLanguages: boolean;
  hasAccessToSensitiveData: boolean;
  fallbackPermissions: string[];
  permissions: CurrentUserResponseModelPermissionsItem[];
  allowedSections: string[];
  isAdmin: boolean;
}
