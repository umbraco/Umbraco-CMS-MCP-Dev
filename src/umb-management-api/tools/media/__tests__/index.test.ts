import { sections } from "@/helpers/umbraco-auth-policies.js";
import { MediaTools } from "../index.js";
import { CurrentUserResponseModel } from "@/umb-management-api/schemas/currentUserResponseModel.js";

describe("media-tool-index", () => {
    it("should have basic query tool by default", () => {
        const userMock = {
            allowedSections: []
        } as Partial<CurrentUserResponseModel>;

        const tools = MediaTools(userMock as CurrentUserResponseModel);
        expect(tools.map(t => t.name)).toMatchSnapshot();
    });

    it("should have tree tools when user has media tree access", () => {
        const userMock = {
            allowedSections: [sections.media]
        } as Partial<CurrentUserResponseModel>;

        const tools = MediaTools(userMock as CurrentUserResponseModel);
        expect(tools.map(t => t.name)).toMatchSnapshot();
    });

    it("should have management tools when user has media section access", () => {
        const userMock = {
            allowedSections: [sections.media]
        } as Partial<CurrentUserResponseModel>;

        const tools = MediaTools(userMock as CurrentUserResponseModel);
        expect(tools.map(t => t.name)).toMatchSnapshot();
    });

    it("should have all tools when user has all required access", () => {
        const userMock = {
            allowedSections: [sections.media]
        } as Partial<CurrentUserResponseModel>;

        const tools = MediaTools(userMock as CurrentUserResponseModel);
        expect(tools.map(t => t.name)).toMatchSnapshot();
    });
});
