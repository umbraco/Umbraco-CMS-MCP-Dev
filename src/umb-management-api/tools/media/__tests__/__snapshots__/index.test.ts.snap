// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`media-tool-index should have all tools when user has all required access 1`] = `
[
  "get-media-by-id",
  "get-media-ancestors",
  "get-media-children",
  "get-media-root",
  "create-media",
  "delete-media",
  "update-media",
  "get-media-configuration",
  "get-media-urls",
  "validate-media",
  "sort-media",
  "get-media-by-id-array",
  "move-media",
  "get-media-audit-log",
  "get-recycle-bin-media-root",
  "get-recycle-bin-media-children",
  "empty-media-recycle-bin",
  "restore-media-from-recycle-bin",
  "move-media-to-recycle-bin",
  "delete-media-from-recycle-bin",
]
`;

exports[`media-tool-index should have basic query tool by default 1`] = `
[
  "get-media-by-id",
]
`;

exports[`media-tool-index should have management tools when user has media section access 1`] = `
[
  "get-media-by-id",
  "get-media-ancestors",
  "get-media-children",
  "get-media-root",
  "create-media",
  "delete-media",
  "update-media",
  "get-media-configuration",
  "get-media-urls",
  "validate-media",
  "sort-media",
  "get-media-by-id-array",
  "move-media",
  "get-media-audit-log",
  "get-recycle-bin-media-root",
  "get-recycle-bin-media-children",
  "empty-media-recycle-bin",
  "restore-media-from-recycle-bin",
  "move-media-to-recycle-bin",
  "delete-media-from-recycle-bin",
]
`;

exports[`media-tool-index should have tree tools when user has media tree access 1`] = `
[
  "get-media-by-id",
  "get-media-ancestors",
  "get-media-children",
  "get-media-root",
  "create-media",
  "delete-media",
  "update-media",
  "get-media-configuration",
  "get-media-urls",
  "validate-media",
  "sort-media",
  "get-media-by-id-array",
  "move-media",
  "get-media-audit-log",
  "get-recycle-bin-media-root",
  "get-recycle-bin-media-children",
  "empty-media-recycle-bin",
  "restore-media-from-recycle-bin",
  "move-media-to-recycle-bin",
  "delete-media-from-recycle-bin",
]
`;
